src/hotspot/share/opto/loopnode.cpp

```diff
@@ -4721,102 +4810,121 @@ void PhaseIdealLoop::verify_compare( Node *n, const PhaseIdealLoop *loop_verify,
+    assert(child->_parent == this, "all must be children of this");
+    children.push(child);
+    child = child->_next;
   }
+}
 
+int compare_tree(IdealLoopTree** a, IdealLoopTree** b) {
+  assert((*a) != nullptr && (*b) != nullptr, "must be");
+  return (*a)->_head->_idx - (*b)->_head->_idx;
 }
 
 //------------------------------verify_tree------------------------------------
-// Verify that tree structures match.  Because the CFG can change, siblings
-// within the loop tree can be reordered.  We attempt to deal with that by
+// Verify that tree structures match. Because the CFG can change, siblings
+// within the loop tree can be reordered. We attempt to deal with that by
 // reordering the verify's loop tree if possible.
-void IdealLoopTree::verify_tree(IdealLoopTree *loop, const IdealLoopTree *parent) const {
-  assert( _parent == parent, "Badly formed loop tree" );
-
-  // Siblings not in same order?  Attempt to re-order.
-  if( _head != loop->_head ) {
-    // Find _next pointer to update
-    IdealLoopTree **pp = &loop->_parent->_child;
-    while( *pp != loop )
-      pp = &((*pp)->_next);
-    // Find proper sibling to be next
-    IdealLoopTree **nn = &loop->_next;
-    while( (*nn) && (*nn)->_head != _head )
-      nn = &((*nn)->_next);
-
-    // Check for no match.
-    if( !(*nn) ) {
-      // Annoyingly, irreducible loops can pick different headers
-      // after a major_progress operation, so the rest of the loop
-      // tree cannot be matched.
-      if (_irreducible && Compile::current()->major_progress())  return;
-      assert( 0, "failed to match loop tree" );
-    }
-
-    // Move (*nn) to (*pp)
-    IdealLoopTree *hit = *nn;
-    *nn = hit->_next;
-    hit->_next = loop;
-    *pp = loop;
-    loop = hit;
-    // Now try again to verify
-  }
-
-  assert( _head  == loop->_head , "mismatched loop head" );
+bool IdealLoopTree::verify_tree(IdealLoopTree* loop_verify) const {
+  assert(_head == loop_verify->_head, "mismatched loop head");
+  assert(this->_parent != nullptr || this->_next == nullptr, "is_root_loop implies has_no_sibling");
+
+  // Collect the children
+  GrowableArray<IdealLoopTree*> children;
+  GrowableArray<IdealLoopTree*> children_verify;
+  collect_children(children);
+  loop_verify->collect_children(children_verify);
+  children.sort(compare_tree);
+  children_verify.sort(compare_tree);
+
+  bool fail = false;
+
+  // Compare the two children lists
+  for (int i = 0, j = 0; i < children.length() || j < children_verify.length(); ) {
+    IdealLoopTree* child        = nullptr;
+    IdealLoopTree* child_verify = nullptr;
+    // Read from both lists, if possible.
+    if (i < children.length()) {
+      child = children.at(i);
+    }
+    if (j < children_verify.length()) {
+      child_verify = children_verify.at(j);
+    }
+    if (child != nullptr && child_verify != nullptr && child->_head != child_verify->_head) {
+      assert(child->_head->_idx != child_verify->_head->_idx, "is implied");
```

Comment:
Why you need this assert? It duplicate the check you already have.

Reply:
Will remove it.